webots_ros2_driver/src/plugins/static/Ros2Compass.cpp

```diff
@@ -0,0 +1,104 @@
+#include "webots_ros2_driver/plugins/static/Ros2Compass.hpp"
+
+#include <webots/compass.h>
+#include <webots/robot.h>
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+
+namespace webots_ros2_driver {
+  void Ros2Compass::init(webots_ros2_driver::WebotsNode *node, std::unordered_map<std::string, std::string> &parameters) {
+    Ros2SensorPlugin::init(node, parameters);
+    // This parameter is read when loading the URDF file
+    mDeviceName = parameters.count("name") ? parameters["name"] : "compass";
+
+    mCompass = wb_robot_get_device(mDeviceName.c_str());
+    mIsEnabled = false;
+
+    assert(mCompass != 0);
+
+    // Data publisher
+    mDataPublisher =
+      mNode->create_publisher<sensor_msgs::msg::MagneticField>(mTopicName + "/values", rclcpp::SensorDataQoS().reliable());
+    // Services
+    mEnableService = mNode->create_service<webots_ros2_msgs::srv::SetInt>(
+      mTopicName + "/enable", std::bind(&Ros2Compass::enable_callback, this, _1, _2));
+    mDisableService = mNode->create_service<webots_ros2_msgs::srv::GetBool>(
+      mTopicName + "/disable", std::bind(&Ros2Compass::disable_callback, this, _1, _2));
+    mGetSamplingPeriodService = mNode->create_service<webots_ros2_msgs::srv::GetInt>(
+      mTopicName + "/get_sampling_period", std::bind(&Ros2Compass::get_sampling_period_callback, this, _1, _2));
+    mGetLookupTableService = mNode->create_service<webots_ros2_msgs::srv::GetFloatArray>(
+      mTopicName + "/get_lookup_table", std::bind(&Ros2Compass::get_lookup_table_callback, this, _1, _2));
+    RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " initialized!").c_str());
+
+    // Calculate timestep
+    if (mAlwaysOn) {
+      wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " activated!").c_str());
+      mIsEnabled = true;
+    }
+  }
+  void Ros2Compass::step() {
+    if (!preStep())
+      return;
+
+    if (mIsEnabled)
+      publishData();
+
+    if (mAlwaysOn)
+      return;
+
+    mustPublish();
+  }
+  bool Ros2Compass::mustPublish() {
+    // Enable/Disable sensor
+    const bool shouldBeEnabled = mDataPublisher->get_subscription_count() > 0;
+    if (shouldBeEnabled != mIsEnabled) {
+      if (shouldBeEnabled)
+        wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      else
+        wb_compass_disable(mCompass);
+      mIsEnabled = shouldBeEnabled;
+    }
+    return mIsEnabled;
+  }
+  void Ros2Compass::publishData() {
+    const double *compass_values = wb_compass_get_values(mCompass);
+    mDataMessage.header.stamp = mNode->get_clock()->now();
+    mDataMessage.magnetic_field.x = compass_values[0];
+    mDataMessage.magnetic_field.y = compass_values[1];
+    mDataMessage.magnetic_field.z = compass_values[2];
+    mDataPublisher->publish(mDataMessage);
```

Comment:
Sorry for randomly jumping into the PR, but a few notes here.

According to the documentation, these values should be in Tesla units:

> x, y, and z components of the field vector in Tesla
- http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/MagneticField.html

Therefore, we should take the raw value from Webots, calculate the actual value using the lookup table and convert those values to Teslas. As the result, the `get_lookup_table` service is irrelevant.

Other services `enable`, `disable`, and `get_sampling_period` are almost never used in real hardware. Therefore, exposing them could potentially make the sim2real transfer harder (we want to minimize the reality gap).

It would be probably useful to publish other types of messages, e.g.:
https://github.com/clearpathrobotics/imu_compass/blob/master/src/imu_compass.cpp
Maybe `sensor_msgs/Imu` with only the `orientation` field populated. Ideally, we should take a look into a few existing compass interfaces and expose only the most generic part of their ROS API.

Reply:
Thanks for your comments! They are really useful. I now understand why all these services are not included in other plugins. I agree that these services should therefore not be included here either. This also applies to the new Emitter and Receiver nodes. I also agree that we should be consistent with other compass interfaces, like the one you have linked.

Reply:
Actually, the Webots Compass node (and magnetometers in general) return a 3D vector of the North direction. In the interface linked by @lukicdarkoo, the code gets the compass measuremement in a message of type [Vector3Stamped](http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/Vector3Stamped.html). We should use this instead of the MagneticField type.

Reply:
@ygoumaz I would probably find Vector3Stamped more useful than MagneticField. Also, a heading (as Float32) would be quite useful for most use cases.